#139: Assets list component

mihailmarcu opened the pull request:
List component

- Heading
- Search UI
- Search functionality

mihailmarcu committed:
#139: Add scoped styles for assets
- Remove the .sp-denom-marker from the utility as it is used directly in the asset component;

mihailmarcu committed:
Merge branch 'release/1.0.0' into feature/139_assets-list-m-2

mihailmarcu committed:
Merge remote-tracking branch 'origin/release/milestone-2' into feature/139_assets-list-m-2

mihailmarcu committed:
#139: Add search bar for assets

mihailmarcu committed:
#139: Add search functionality

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -57,26 +76,74 @@ export default defineComponent({
       )
     })
 
+    const filteredBalanceList = computed(() => {
```
Since we also have a search going on [here](https://github.com/tendermint/vue/blob/e5bf4b3b7a241835253a13fa26aafbd8ba9451cc/packages/vue/src/components/SpAmountSelect/SpAmountSelect.vue#L182), we might as well come up with a composable, e.g. useSearch... 

Doesn't need to be in this PR though.

mihailmarcu replied:
Good idea. I also thought about a composable for search and we could send parameters like search criteria and more. Let's create a separate task for this later.

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -57,26 +76,74 @@ export default defineComponent({
       await queryAllBalances({
-        params: { address: props.address },
+        params: { address: props.address},
+        query: {"pagination.reverse": false},
```
didn't know about this reverse thing...

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -57,26 +76,74 @@ export default defineComponent({
   }
 })
+
+function sortList(list) {
```
shouldn't sortList be a computed?

mihailmarcu replied:
This is the main logic that will probably be changed.

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -57,26 +76,74 @@ export default defineComponent({
+      searchString = searchString.trim().toLowerCase()
+
+      searchArray = searchArray.filter((item) => {
+        if (item.denom.toLowerCase().indexOf(searchString) !== -1) {
```
maybe it's better to use `.includes` so to not depend on whole words?

mihailmarcu replied:
Working on refactoring this functionality.

marinhoarthur replied:
shall we use includes?

marinhoarthur replied:
correction: `indexOf` works just fine, i.e. it is as capable as `includes`

mihailmarcu replied:
I will use `includes` as it looks more clear.

mihailmarcu committed:
#139: Add show more functionality and demo data

marinhoarthur commented on packages/vue/tsconfig.json:
```diff
@@ -18,7 +18,8 @@
     "paths": {
-      "@/*": ["src/*"]
+      "@/*": ["src/*"],
+      "#/types": ["src/utils/types"]
```
nice

mihailmarcu replied:
This was accidentally pushed.

mihailmarcu committed:
#139: Remove the sort of the assets in case of the non search value

mihailmarcu committed:
#139: Removed sort functionality as it's not needed

mihailmarcu committed:
Merge remote-tracking branch 'origin/feature/sptx-with-ibc' into feature/139_assets-list-m-2

mihailmarcu committed:
#139: Add size prop for SpDenom
- Use SpDenom for assets list
- resultLimit is always a number
- concat the balances with data when is not empty

mihailmarcu committed:
#139: Reset the displayLimit after remove search criteria

mihailmarcu committed:
#139: Add no result state
- Improve the logic when the show more is displayed

mihailmarcu committed:
#139: Add no assets state
- Add skeleton animation

mihailmarcu committed:
#139: Refactor size prop for SpDenom

mihailmarcu committed:
#139: Improve UI and UX for search

mihailmarcu committed:
#139: Add arrow for show more button

fadeev commented:
@mihailmarcu please, resolve conflicts.

@marinhoarthur please, review and merge.

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -22,31 +48,125 @@
+    const searchInput = ref<null | { focus: () => null }>(null)
+
+    // state
+    let uiState: State = reactive(initialState)
```
why two different state objects?

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -22,31 +48,125 @@
+      searchQuery: "",
+      displayLimit: props.resultLimit,
+      demoDenoms:[
+        {
```
let's don't forget to clean up dummy data

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -22,31 +48,125 @@
   setup(props) {
     // store
     const $s = useStore()
+    const searchInput = ref<null | { focus: () => null }>(null)
```
I think we can go for a single state object

mihailmarcu replied:
This is how it was intended, TS complains about focus(). I didn't find any other solution to solve this problem other than to create another ref value.

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -2,17 +2,43 @@
+              </Suspense>
             </div>
             <div class="sp-denom-name">
               {{ balance.denom.toUpperCase() }}
```
let's use SpDenom everywhere a denom appears

mihailmarcu replied:
Absolutely agree.

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -57,26 +177,155 @@ export default defineComponent({
+          return item
+        }
+      })
+      filteredArrayLength = searchArray.length
```
why do we need this?

`isShowMore` and `filteredBalanceList` could use some simplification

mihailmarcu replied:
This logic is a bit confusing, I agree. If you have any ideas how it can be simplified then I will gladly do it.

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -57,26 +177,155 @@ export default defineComponent({
+        query: {"pagination.reverse": false},
         options: { all: true, subscribe: true }
+      }).then(() => {
+        setTimeout(() => uiState.isAssetsLoading = false, 1500);
```
to remove

mihailmarcu replied:
The timeout is removed, but the callback must be left to change the loading state.

mihailmarcu committed:
#139: Use SpDenom for IBC

mihailmarcu committed:
#139: Add fixes after PR review

mihailmarcu committed:
#139: Improve the loading state

marinhoarthur commented on packages/vue/src/components/SpAssets/SpAssets.vue:
```diff
@@ -57,26 +133,163 @@ export default defineComponent({
+      }
+    })
+
+    const filteredBalanceList = computed(() => {
```
this is still over complicated =

- filtered computed should only care about filtering 
- no need to create local vars from outer ones
- no need to care for balances being empty (see comp. getter)
- no need to store length of filtered arr

mihailmarcu committed:
#139: WIP for refactoring the search functionality

mihailmarcu committed:
Merge branch 'release/milestone-2' into feature/139_assets-list-m-2

# Conflicts:
#	packages/vue/src/components/SpDenom/SpDenom.vue
#	packages/vue/src/composables/useDenom.ts

mihailmarcu committed:
Merge branch 'release/milestone-2' into feature/139_assets-list-m-2

# Conflicts:
#	packages/vue/src/components/SpAssets/SpAssets.vue
#	packages/vue/src/components/SpDenom/SpDenom.vue

mihailmarcu committed:
#139: Refactoring and improvements for assets component

mihailmarcu committed:
Merge branch 'release/milestone-2' into feature/139_assets-list-m-2

# Conflicts:
#	packages/vue/src/components/SpAmountSelect/SpAmountSelect.vue
#	packages/vue/src/components/SpDenom/SpDenom.vue
#	packages/vue/src/components/SpTx/SpTx.vue

mihailmarcu committed:
#139: Minor adjustments for Tx component

mihailmarcu committed:
#139: Add reset search functionality

mihailmarcu committed:
#139: Remove log helpers
